File: LeetCode/133. Clone Graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return node;
        Map<Node, Node> map = new HashMap<>(); //old node: new node
        
        class H {
            Node dfs(Node oldNode) {
                if (oldNode == null) return oldNode;
                if (map.containsKey(oldNode)) {
                    return map.get(oldNode);
                }
                Node newNode = new Node(oldNode.val);
                map.put(oldNode, newNode);

                for (Node node : oldNode.neighbors) {
                    newNode.neighbors.add(dfs(node));
                }
                return newNode;
            }
        }


        return new H().dfs(node);
    }
} //TC: O(n+e) SC: O(n), where n is the node count, e is the edge count

File: LeetCode/133. Clone Graph.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var neighbors: ArrayList<Node?> = ArrayList<Node?>()
 * }
 */

class Solution {
    val newMap = HashMap<Node, Node>() //old node:new node
    fun cloneGraph(node: Node?): Node? {
        return dfs(node)
    }

    fun dfs(node: Node?): Node? {
        if (node != null) {
            if(node in newMap) return newMap[node] //if node has the same one in the map, return the copied node

            val cp = Node(node.`val`) //copy a new node
            newMap[node] = cp //if not, add the node and link the newly copied node with the old node
            for (n in node.neighbors) { //loop the neighbor nodes of the old node
                cp.neighbors.add(dfs(n)) //add copied neighbors to the copy node
            }
            return cp
        }
        return null //do nothing and return null if the node is null
    }
}
